src/fsharp/service/ServiceUntypedParse.fs

```diff
@@ -1283,6 +1282,47 @@ module UntypedParseImpl =
             }
 
         AstTraversal.Traverse(pos, pt, walker)
+        // Uncompleted attribute applications are not presented in the AST in any way. So, we have to parse source string.
```

Comment:
My gut feeling tells me we should try to get something useful from the parse tree instead of trying to do some sort of ad-hoc parsing ourselves to determine what context we are in.

Consider this code:
```fsharp
[<
 (* marker *)
 >]
module Test =

    let beef = 1
```

Completions do not work here because we are checking `lineStr`. IMHO, we should avoid `lineStr` and parsing ourselves to determine what context we are in at all costs.

If we are getting nothing useful from the parse tree, then we need to modify the parser. That way we can handle most if not all cases. The implementation would be much simpler too.

Also, in the future, these big new lines code should probably go into a separate function rather than make an existing function bigger than what it was.

Reply:
@TIHan This also applies for current uses of QuickParse that uses a similar approach of getting a context for possibly unfinished expressions. Moving to AST there as well would be a great move.